internal/security/acctz/acctz.go

```diff
@@ -0,0 +1,917 @@
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+const (
+	successUsername      = "acctztestuser"
+	successPassword      = "verysecurepassword"
+	failUsername         = "bilbo"
+	failPassword         = "baggins"
+	failRoleName         = "acctz-fp-test-fail"
+	successCliCommand    = "show version"
+	failCliCommand       = "show version"
+	shellCommand         = "uname -a"
+	gnmiCapabilitiesPath = "/gnmi.gNMI/Capabilities"
+	gnoiPingPath         = "/gnoi.system.System/Ping"
+	gnsiGetPath          = "/gnsi.authz.v1.Authz/Get"
+	defaultSSHPort       = 22
+)
+
+var gRPCClientAddr net.Addr
+
+// Record represents the structure for an acctz record.
+type Record struct {
+	StartTime            time.Time
+	DoneTime             time.Time
+	CmdType              acctz.CommandService_CmdServiceType
+	Cmd                  string
+	RPCType              acctz.GrpcService_GrpcServiceType
+	RPCPath              string
+	RPCPayload           string
+	LocalIP              string
+	LocalPort            uint32
+	RemoteIP             string
+	RemotePort           uint32
+	Succeeded            bool
+	ExpectedStatus       acctz.SessionInfo_SessionStatus
+	ExpectedAuthenType   acctz.AuthnDetail_AuthnType
+	ExpectedAuthenStatus acctz.AuthnDetail_AuthnStatus
+	ExpectedAuthenCause  string
+	ExpectedIdentity     string
+}
+
+func setupUserPassword(t *testing.T, dut *ondatra.DUTDevice, username, password string) {
+	request := &credentialz.RotateAccountCredentialsRequest{
+		Request: &credentialz.RotateAccountCredentialsRequest_Password{
+			Password: &credentialz.PasswordRequest{
+				Accounts: []*credentialz.PasswordRequest_Account{
+					{
+						Account: username,
+						Password: &credentialz.PasswordRequest_Password{
+							Value: &credentialz.PasswordRequest_Password_Plaintext{
+								Plaintext: password,
+							},
+						},
+						Version:   "v1.0",
+						CreatedOn: uint64(time.Now().Unix()),
+					},
+				},
+			},
+		},
+	}
+
+	credzClient := dut.RawAPIs().GNSI(t).Credentialz()
+	credzRotateClient, err := credzClient.RotateAccountCredentials(context.Background())
+	if err != nil {
+		t.Fatalf("Failed fetching credentialz rotate account credentials client, error: %s", err)
+	}
+	err = credzRotateClient.Send(request)
+	if err != nil {
+		t.Fatalf("Failed sending credentialz rotate account credentials request, error: %s", err)
+	}
+	_, err = credzRotateClient.Recv()
+	if err != nil {
+		t.Fatalf("Failed receiving credentialz rotate account credentials response, error: %s", err)
+	}
+	err = credzRotateClient.Send(&credentialz.RotateAccountCredentialsRequest{
+		Request: &credentialz.RotateAccountCredentialsRequest_Finalize{
+			Finalize: request.GetFinalize(),
+		},
+	})
+	if err != nil {
+		t.Fatalf("Failed sending credentialz rotate account credentials finalize request, error: %s", err)
+	}
+
+	// Brief sleep for finalize to get processed.
+	time.Sleep(time.Second)
+}
+
+func nokiaFailCliRole(t *testing.T) *gnmi.SetRequest {
```

Comment:
is nokia somehow special? (why call this 'nokiaFail....' ?)

Reply:
This is for cli tests where user needs to send some cli commands which are denied. Currently it seems there are no corresponding OC paths to define denied commands for users. So each vendor would need to create a failure cli role (roles with certain commands not allowed) in their native model.